docs/source/proposals/jit-classes.rst

```diff
@@ -0,0 +1,200 @@
+
+
+Reference count and destructor
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An instance of jit-class is reference-counted by NRT. Since it may contain
+other NRT tracked object, it must call a destructor when its reference count
+dropped to zero.  The destructor will decrement the reference count of all
+attributes by one.
+
+At this time, there is no support for user defined ``__del__`` method.
+
+Cyclic reference is not handled at this time.  It will cause memory leak.
+
+Type inference
+~~~~~~~~~~~~~~
+
+So far we have not described the type of the attributes or the methods.
+Type information is necessary to materailize the instance (e.g. allocate the
+storage).  The simplest way is to let user provide the type of each attributes
+as well as the ordering; for instance::
+
+    dct = OrderedDict()
+    dct['x'] = int32
+    dct['y'] = float32
+
+Allowing user to supply an ordered dictionary will provide the name, ordering
+and types of the attributes.  However, this statically typed semantic is not as
+flexible as the Python semantic which behaves like a generic class.
+
+Inferring the type of attributes is difficult.  In a previous attempt to
+implement JIT classes, the ``__init__`` method is specialized to capture
+the type stored into the attributes.  Since the method can contain arbitrary
+logic, the problem can become a dependent typing problem if types are assigned
+conditionally depending on the value. (Very few languages implement dependent
+typing and those that does are mostly theorem provers.)
+
+Example: typing function using an OrderedDict
+---------------------------------------------
+
+.. code-block:: python
+
+    spec = OrderedDict()
+    spec['x'] = numba.int32
+    spec['y'] = numba.float32
+
+    @jitclass(spec)
+    class Vec(object):
+        def __init__(self, x, y):
+            self.x = x
+            self.y = y
+
+        def add(self, dx, dy):
+            self.x += dx
+            self.y += dy
+
+Example: typing function using an list of 2-tuples
+--------------------------------------------------
+
+.. code-block:: python
+
+    spec = [('x', numba.int32),
+            ('y', numba.float32)]
+
+    @jitclass(spec)
+    class Vec(object):
+        ...
+
+Usage from the Interpreter
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When constructing a new instance of a jitclass, a "box" is created that wraps
+the underlying jitclass instance from numba.  Attributes and methods are
+accessible from the interpreter.  The actual implementation will be in numba
+compiled code.  Any Python object is converted to its native
+representation for consumption in numba.  Similarly, the returned value is
+converted to its Python representation.  As a result, there may be overhead
+in manipulating jitclass instances in the interpreter.
+
+Support for property, staticmethod and classmethod
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The use of ``property`` is accepted for getter and setter only.  Delter is not
```

Comment:
Not yet, that's something I meant to add in https://github.com/numba/numba/pull/1571